lib/di/interceptor/app_interceptor.dart

```diff
@@ -28,48 +31,59 @@ class AppInterceptor extends Interceptor {
       options.headers
           .putIfAbsent(_authorizationHeader, () => '$tokenType $accessToken');
     }
-    return super.onRequest(options, handler);
+    super.onRequest(options, handler);
   }
 
   @override
   void onError(DioError err, ErrorInterceptorHandler handler) {
-    handler.next(err);
-    // TODO: Integrate refresh-token https://github.com/nimblehq/ic-flutter-taher-toby/issues/20
-    // final statusCode = err.response?.statusCode;
-    // if ((statusCode == HttpStatus.forbidden ||
-    //         statusCode == HttpStatus.unauthorized) &&
-    //     _requireAuthentication) {
-    //   _doRefreshToken(err, handler);
-    // } else {
-    // }
+    final statusCode = err.response?.statusCode;
+    if ((statusCode == HttpStatus.forbidden ||
+            statusCode == HttpStatus.unauthorized) &&
+        _requireAuthentication) {
+      _doRefreshToken(err, handler);
+    } else {
+      handler.next(err);
+    }
   }
 
-  // ignore: unused_element
   Future<void> _doRefreshToken(
-    DioError err,
+    DioError error,
     ErrorInterceptorHandler handler,
   ) async {
     try {
-      // if (result is Success) {
-      // err.requestOptions.headers[_headerAuthorization] = newToken;
-      // Create request with new access token
-      final options = Options(
-          method: err.requestOptions.method,
-          headers: err.requestOptions.headers);
-      final newRequest = await _dio.request(
-          "${err.requestOptions.baseUrl}${err.requestOptions.path}",
+      final RefreshTokenUseCase refreshTokenUseCase =
+          getIt.get<RefreshTokenUseCase>();
+      final StoreAuthTokenUseCase storeAuthTokenUseCase =
+          getIt.get<StoreAuthTokenUseCase>();
+      final String refreshToken =
+          await _secureStorage.readSecureData(refreshTokenKey) ?? '';
+      final refreshTokenResult = await refreshTokenUseCase.call(refreshToken);
+      if (refreshTokenResult is Success<AuthTokenModel>) {
+        AuthTokenModel authTokenModel = refreshTokenResult.value;
+        final String accessToken = authTokenModel.accessToken;
+        final String tokenType = authTokenModel.tokenType;
+        storeAuthTokenUseCase.call(authTokenModel);
```

Comment:
So with this logic, we're saving the tokens by a use case, but getting them back directly from the `SecureStorage` component? Is that our intention or something we would like to improve as I don't see any FIXME comment here 💭

Reply:
@luongvo @Tuubz  I am a bit confused here, should we create a new usecase class for getting/reading the saved token data?

Reply:
I think now it's up to you guys here. `AppInterceptor` is a special component, so using `SecuredStorage` to get the simple token data is not so bad. I can skip this review 👍